goship.go

```diff
@@ -760,23 +791,216 @@ func extractOutputHandler(fn func(http.ResponseWriter, *http.Request, string, st
+	if err != nil {
+		panic(err)
+	}
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		state := gomniauth.NewState("after", "success")
+
+		authUrl, err := provider.GetBeginAuthURL(state, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		http.Redirect(w, r, authUrl, http.StatusFound)
+	}
+}
+
+func callbackHandler(providerName string, auth bool) http.HandlerFunc {
+
+	if auth != true {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+
+	provider, err := gomniauth.Provider(providerName)
+	if err != nil {
+		panic(err)
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		omap, err := objx.FromURLQuery(r.URL.RawQuery)
+		if err != nil {
+			log.Printf("error getting resp from callback")
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		creds, err := provider.CompleteAuth(omap)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		user, userErr := provider.GetUser(creds)
+		if userErr != nil {
+			log.Printf("Failed to get user from Github %s", user)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		session, err := store.Get(r, sessionName)
+		if err != nil {
+			log.Fatal("Can't get a session store")
+		}
+
+		session.Options = &sessions.Options{
+			Path:     "/",
+			MaxAge:   86400 * 7,
+			HttpOnly: true,
+		}
+
+		session.Values["userName"] = user.Nickname()
+		session.Values["avatarURL"] = user.AvatarURL()
+		session.Save(r, w)
+
+		http.Redirect(w, r, os.Getenv("GITHUB_CALLBACK_URL"), http.StatusFound)
+	}
+}
+
+type auth struct {
+	authorization       bool
+	githubRandomHashKey string
+	githubOmniauthID    string
+	githubOmniauthKey   string
+	githubCallbackURL   string
+}
+
+//  Authenticate with Github. If env data is missing turn Auth off.
+func getAuth() auth {
+	a := auth{}
+	a.authorization = true
+	a.githubRandomHashKey = os.Getenv("GITHUB_RANDOM_HASH_KEY")
+	a.githubOmniauthID = os.Getenv("GITHUB_OMNI_AUTH_ID")
+	a.githubOmniauthKey = os.Getenv("GITHUB_OMNI_AUTH_KEY")
+	a.githubCallbackURL = os.Getenv("GITHUB_CALLBACK_URL")
+	// Let user know if a key is missing and that auth is disabled.
+	if a.githubRandomHashKey == "" || a.githubOmniauthID == "" || a.githubOmniauthKey == "" || a.githubCallbackURL == "" {
+		log.Printf("Missing one or more Gomniauth Environment Variables: Running with with limited functionality! \n githubRandomHashKey [%s] \n githubOmniauthID [%s] \n githubOmniauthKey[%s] \n githubCallbackURL[%s]",
+			a.githubRandomHashKey,
+			a.githubOmniauthID,
+			a.githubOmniauthKey,
+			a.githubCallbackURL)
+		a.authorization = false
+	} else {
+		gomniauth.SetSecurityKey(a.githubRandomHashKey)
+		gomniauth.WithProviders(
+			githubOauth.New(a.githubOmniauthID, a.githubOmniauthKey, a.githubCallbackURL+"/auth/github/callback"),
+		)
+		a.authorization = true
+	}
+	return a
```

Comment:
The point is not so much about how many return statement or exit point a function should have. But about code readability.

Code readability is the most important thing.

If the code is more readable with more that one exit point than that is the way it should be done.
However if a peace of code turns out to be more readable with only one exit point at the end than that is the way it should be.